### DISCUSSION TITLE
Neue CI Pipeline

### DISCUSSION THREAD

**Discussion by psi-4ward** [pa-pa/AskSinPP#236] [Ideas]
Hi,
da Travis-CI zunehmend [Probleme](https://github.com/pa-pa/AskSinPP/issues/234#issuecomment-776944215) macht und die Erfahrungen mit Github-Actions meiners doch recht positiv sind war die Überlegung umzusteigen. In diesem Zug würde ich auch die CI etwas anders aufbauen und wollte dies mit euch mal diskutieren.

Zunächst scheint die Arduino-IDE mittlerweile auch auf der [arduino-cli aufzubauen](https://blog.arduino.cc/2020/03/13/arduino-cli-an-introduction/) weshalb ich diese nutzen würde.

> The needs of our community, as well as our own, prompted us to radically rethink our approach and embark on a major refactoring operation with the goal to obtain a single reference implementation: the Arduino Command Line Interface.

> As we move forward, all our legacy products are being reworked to use the Arduino CLI as the foundation to build upon.

Zudem würde ich die ganzen Steps in Bash-Scripts gießen, damit ist man n gutes Stück unabhängig vom CI-Runner und kann es auch lokal laufen lassen. (Wie wir das für Windows umsetzen müsste man sich nicht überlegen - hatte mal gehört Windows kann jetzt auch Bash?)

Bezüglich externer Libs: Hier hatte ich mit @jp112sdl schon gesprochen und der Konsens war, wie bisher _latest_ aus den Git-Repos zu ziehen.

**Comment by pa-pa** [comment]
Ich möchte eigentlich keine bash-Scripte lokal ausführen, in die andere Leute einfach was einchecken können :-(

**Reply by psi-4ward** [reply]
Hmm, wie meinst du das?

**Reply by pa-pa** [reply]
Wenn in den Bash-Scripten zu viel Magie - kann später mal ein "böswilliger" auch mal ein "rm -rf /" irgendwo einbauen. Auf dem GitHub-CI-System ist mir das egeal - aber auf meine System will ich sowas nicht erleben. Hier sieht man gerade schön, was alles mit automatischen Scripting beim CI möglich ist - https://www.heise.de/news/Sicherheitsforscher-bricht-ueber-Open-Source-Repositories-bei-PayPal-Co-ein-5051635.html?wt_mc=rss.red.ho.ho.atom.beitrag.beitrag

**Reply by psi-4ward** [reply]
Nuja dann musst du den Schadcode aber erst mal mergen und dein System das zulassen nachdem du es selbst explizit aufgerufen hast. Ich seh da bei uns ein sehr sehr geringens Risiko

**Comment by psi-4ward** [comment]
Wie wäre es damit:
* https://github.com/psi-4ward/AskSinPP/runs/1880752379
* https://github.com/psi-4ward/AskSinPP/tree/master/tests

Das könnte man auch schnell lokal laufen lassen und wenn wir in 2 Jahren alle auf GitLab migrieren weil aus Gründen dann kann man die Scripts hier genau so verwenden. Natürlich könnte man hier und da noch etwas optimieren.
Auch wenn noch nicht alles Sketches drin sind ist die Laufzeit von 3min doch ganz ok.

**Reply by jp112sdl** [reply]
Das sieht wirklich gut aus

**Reply by psi-4ward** [reply]
Wenn die sketches in Unterordnern nach board wären könnte man nach ino Dateien suchen und müsste nicht explizit auflisten. Wäre auch für die Bastler ne ganz nette Info

**Reply by pa-pa** [reply]
Sieht cool - könnte man nicht die platformio.ini Files für die Boardauswahl nutzen ?
Ich würde auf das "automatische" Sammeln der Sketches verzichten, da man dann keinen Sketch ausnehmen kann. Bedeutet zwar etwas mehr Handarbeit - aber so oft kommen ja keine Sketches dazu. Feste Strukturen ist auch eher schwer, da dann die Flexibilität verloren geht.
Ich träume auch immer noch von einer Aufzeichnung der Größe der gebauten Sketches, um einfacher gute und schlechte Änderungen in der Implementierung erkennen kann. Aber das hat jetzt nicht Priorität.

**Reply by psi-4ward** [reply]
Natürlich könnte man die platform.ini dazu nutzen aber dann muss jeder Sketch eine solche haben.

Eine Alternative wäre eine Kommentarzeile die ausgewertet wird:

´´´cpp
// ci-test=yes board=arduino:avr:pro:cpu=8MHzatmega328 aes=yes
#include <AskSinPP.h>
...
´´´

Gebaut wären beide Varianten relativ schnell.

**Reply by jp112sdl** [reply]
Ich wäre für die Variante `// ci-test=...` als Kommentar im Sketch

**Reply by pa-pa** [reply]
Den Kommentar finde ich cool. Dann geht auch automatisch allea aufsammeln :-)

**Reply by psi-4ward** [reply]
Dann bau ich das so ;)

**Comment by psi-4ward** [comment]
https://github.com/psi-4ward/AskSinPP/actions/runs/559147101

Es wird ;)

Allerdings baut der STM32 nicht wenn ich `build.extra_flags` für AES mitgebe. Da müssten wir noch mal schauen

**Reply by jp112sdl** [reply]
AES wird aber auch aktuell bei Travis nicht beim STM32 benutzt, wenn ich das richtig sehe.
Der extra_flag hier
https://github.com/pa-pa/AskSinPP/blob/69251434746c95091a403a28d16f2eb8aaa3dcea/.travis.yml#L81
ist nur für `/hardware/arduino/avr`

**Reply by psi-4ward** [reply]
Das ist richtig aber eigentlich müsste es ja gehen...

**Reply by jp112sdl** [reply]
Warum müsste es gehen?

In der `aes.h` steht `AVR-Crypto-Lib.`

**Reply by psi-4ward** [reply]
Also es kompiliert mit `compiler.cpp.extra_flags und der Speicherverbrauch wird etwas höher.

**Reply by psi-4ward** [reply]
https://github.com/psi-4ward/AskSinPP/actions/runs/559231743

ez schauts gut aus. Man beachte auch die Annotations unter "Summary" :)

**Reply by jp112sdl** [reply]
Cool wäre eine Größendifferenz je Sketch zu bekommen... aber das wird dann auch unübersichtlich.
Weiß nicht, ob der Durchschnitt viel Nährwert hat. Vielleicht um eine Tendenz zu erkennen.

Irritierend ist ein wenig, dass da "warning" steht

**Reply by psi-4ward** [reply]
Ja gibt zumindest mal was aus. "Info" kann man leider nicht ohne API Call erzeugen - gibt n altes Issue dazu.
Wenn das "wichtig" ist könnte man da natürlich noch Zeit investieren.

**Reply by pa-pa** [reply]
Das sieht doch schon mal sehr gut aus. Danke

**Reply by jp112sdl** [reply]
Find ich auch!
Danke für die Mühe. 

Ein paar Kleinigkeiten sind noch zu klären.

1.) Aus der Projekt-Liste https://github.com/pa-pa/AskSinPP/blob/master/.github/workflows/tests/download-sketches.sh wird nur `"https://github.com/jp112sdl/Beispiel_AskSinPP.git"` geladen. Der Rest nicht

2.) Bei den 1284P Boards gibt es Sketche mit BOBUINO Layout und welche mit STANDARD Layout. Hab das erstmal mit 2 Boards gelöst... 1284pBOB und 1284pSTD. Vielleicht kann man das noch eleganter über einen `//ci-test layout=bobuino` Kommentar lösen

**Reply by psi-4ward** [reply]
> Find ich auch!
> Danke für die Mühe.
> 
> Ein paar Kleinigkeiten sind noch zu klären.
> 
> 1.) Aus der Projekt-Liste https://github.com/pa-pa/AskSinPP/blob/master/.github/workflows/tests/download-sketches.sh wird nur `"https://github.com/jp112sdl/Beispiel_AskSinPP.git"` geladen. Der Rest nicht

Offensichtlich ein Bug ;) Schau ich mir an.

> 2.) Bei den 1284P Boards gibt es Sketche mit BOBUINO Layout und welche mit STANDARD Layout. Hab das erstmal mit 2 Boards gelöst... 1284pBOB und 1284pSTD. Vielleicht kann man das noch eleganter über einen `//ci-test layout=bobuino` Kommentar lösen

Wenn wir `layout=xyz` einbauen brauchen wir wohl doch noch ein Mapping auf den FQBN. Denk mal die Variante mit `1284p-bob` ist hier einfacher.

**Comment by psi-4ward** [comment]
@pa-pa 
Was kommt in das `Sketch` Verzeichnis?
https://github.com/pa-pa/AskSinPP/commit/af4a869f3ec1358353222596513ea04636305e3f

**Reply by pa-pa** [reply]
https://github.com/pa-pa/HB-Sec-RHS-3/tree/master/Sketch

Error: error sketches/HB-Sec-RHS-3/Sketch/HB-SEC-RHS-3.ino

Das liegt an der blöden Arduino-CLI :-(

**Reply by jp112sdl** [reply]
Ja, Sketch und Ordner müssen gleich heißen

**Reply by pa-pa** [reply]
Macht im HB-Sec-RHS-3 Repo natürlich wenig Sinn :-(

**Reply by jp112sdl** [reply]
Du brauchst die .ino ja nur aus dem Sketch Ordner ins Hauptverzeichnis schieben, dann passt es (Groß-/Kleinschreibung noch anpassen)

Jeder der daheim den Sketch mit der Arduino IDE aus dem geklonten Repo kompilieren will, bekommt ohnehin immer eine Fehlermeldung und muss den Sketch verschieben.

**Reply by jp112sdl** [reply]
Oder du benennst den Sketch in "Sketch.ino" um :)

**Reply by pa-pa** [reply]
ne - das soll extra - dann eher noch ein extra Verzeichnis

**Comment by psi-4ward** [comment]
Ist die Unterscheidung zw AES und nicht-AES bei der Byte-Zählung relevant @pa-pa ?

**Reply by pa-pa** [reply]
nö
